src/bucket/LedgerCmp.h

```diff
@@ -88,62 +91,198 @@ struct LedgerEntryIdCmp
+            // Entries are identical
+            return false;
+        case TRUSTLINE:
+            return lexCompare(aid, bid, a.trustLine().asset,
+                              b.trustLine().asset);
+        case OFFER:
+            return lexCompare(aid, bid, a.offer().offerID, b.offer().offerID);
+        case DATA:
+            return lexCompare(aid, bid, a.data().dataName, b.data().dataName);
+        default:
+            throw std::runtime_error("Unexpected entry type.");
+        }
+        return false;
+    }
+};
+
+namespace
+{
+
+// Compares BucketEntry's using given LedgerCmp function.
+template <class LedgerCmp>
+bool
+bucketEntryIdCmpInternal(BucketEntry const& a, BucketEntry const& b)
+{
+    BucketEntryType aty = a.type();
+    BucketEntryType bty = b.type();
+
+    // METAENTRY sorts below all other entries, comes first in buckets.
+    if (aty == METAENTRY || bty == METAENTRY)
+    {
+        return aty < bty;
+    }
+
+    if (aty == LIVEENTRY || aty == INITENTRY)
+    {
+        if (bty == LIVEENTRY || bty == INITENTRY)
+        {
+            return LedgerCmp{}(a.liveEntry().data, b.liveEntry().data);
         }
         else
         {
-            if (aty != DEADENTRY)
-            {
-                throw std::runtime_error(
-                    "Malformed bucket: unexpected non-INIT/LIVE/DEAD entry.");
-            }
-            if (bty == LIVEENTRY || bty == INITENTRY)
+            if (bty != DEADENTRY)
             {
-                return LedgerEntryIdCmp{}(a.deadEntry(), b.liveEntry().data);
+                throw std::runtime_error("Malformed bucket: unexpected "
+                                         "non-INIT/LIVE/DEAD entry.");
             }
-            else
+            return LedgerCmp{}(a.liveEntry().data, b.deadEntry());
+        }
+    }
+    else
+    {
+        if (aty != DEADENTRY)
+        {
+            throw std::runtime_error(
+                "Malformed bucket: unexpected non-INIT/LIVE/DEAD entry.");
+        }
+        if (bty == LIVEENTRY || bty == INITENTRY)
+        {
+            return LedgerCmp{}(a.deadEntry(), b.liveEntry().data);
+        }
+        else
+        {
+            if (bty != DEADENTRY)
             {
-                if (bty != DEADENTRY)
-                {
-                    throw std::runtime_error("Malformed bucket: unexpected "
-                                             "non-INIT/LIVE/DEAD entry.");
-                }
-                return LedgerEntryIdCmp{}(a.deadEntry(), b.deadEntry());
+                throw std::runtime_error("Malformed bucket: unexpected "
+                                         "non-INIT/LIVE/DEAD entry.");
             }
+            return LedgerCmp{}(a.deadEntry(), b.deadEntry());
         }
     }
+}
+}
+
+/**
+ * Compare two BucketEntries for identity by comparing their respective
+ * LedgerEntries (ignoring their hashes, as the LedgerEntryCmp ignores their
+ * bodies).
+ */
+struct BucketEntryIdCmp
+{
+    virtual bool
```

Comment:
I'm still not sure the best way to go about the compare functions. I need the two compare functions to have the same type to be able to pass them as parameters to functions. To avoid class inheritance and virtual dispatch, I ended up with a `typedef` of the compare function prototype with `std::function`, and changed the two compare functors to be a single, templated function. The template allows you to specify an additional type parameter while still maintaining the `std::less` prototype, and the `typedef` lets you pass the functions as a parameter. It works and seems syntactically sound, but I still feel like it's not the best interface.

Reply:
Yeah, reading it I'm not sure it's a net improvement or loss, or if it'd help to make things even-more-templated / statically type-based; but there's no point going around in circles on it, I'm fine with this version for now.